### PR TITLE
Ac initial

### PR DESCRIPTION
**Pull request by txoof**
This should create the insideCuts and outsideCuts

Test with 
`outsideCuts();` 
`insideCuts();`

### CODE REVIEW

File: flat_pack_library.scad [hunk 1]
```diff
@@ -0,0 +1,87 @@
+//openscad flat-pack joint library
+
+
+
+module outsideCuts(length=100, finger=8, material=5, text=false, center=false, font="Liberation Sans") {
```
Comment (xxv) [inline flat_pack_library.scad]: I'm not sure I understand what the purpose of the defaults here is. Is there something special about 10cm, with 8mm fingers and 5mm thickness? These parameters will likely vary with application considerably.

File: flat_pack_library.scad [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  // overage to ensure that all cuts are completed
+  o = 0.0001;
```
Comment (xxv) [inline flat_pack_library.scad]: `o` looks a lot like `0` when skimming code. Perhaps make a more descriptive name?

File: flat_pack_library.scad (hunk 1 above)
Comment (xxv) [reply flat_pack_library.scad]: Also, perhaps a small comment could be placed above the module just mentioning what the parameters are. Is it the length of the fingers or the number of fingers? etc.

File: flat_pack_library.scad [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+  // number of "female cuts"
+  num_cuts = ceil(usable_divsions/2);
+
+  //set position relative to origin
+  x_translation = center==false ? 0 : -usable_divsions*finger/2;
+  y_translation = center==false ? 0 : -material/2;
+
+  translate([x_translation, y_translation]) {
+    for (i=[0:num_cuts-1]) {
```
Comment (xxv) [inline flat_pack_library.scad]: space around the `:`, `=`, and operators may help readability here.

File: flat_pack_library.scad [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        translate([i*finger*2+padding, -o/2])
+          square([end_cut_length, material+o]);
+      }
+    }
+  }
+
+  debugText = finger>=length/3 ? "ERR: finger>1/3 length" : "outsideCut";
+
+
+  if (text) {
```
Comment (xxv) [inline flat_pack_library.scad]: As mentioned in IRC, perhaps there's a better way to add debugging information here. Perhaps you create a debug module that wraps these modules:

```
module outside_cuts_debug(length, finger, material) {
  outside_cuts(length, finger);
  // ...
  translate([length/2+x_translation, y_translation+material*2])
    text(text=debugText, size = length*.05, halign = "center", font = font);
}
```

**Review by xxv — CHANGES_REQUESTED** [review]
Looking good! The main concern I've got is the debug text. Cheers!

File: flat_pack_library.scad (hunk 1 above)
Comment (txoof) [reply flat_pack_library.scad]: The defaults ensure that something is generated similar to the 1x1x1 cube that a `cube();` with no parameters generates. Is this bad practice? The default values are nothing special.

Comment (xxv) [reply flat_pack_library.scad]: Ah, makes sense. In programming generally such defaults are bad, but if other parts of OpenSCAD do it then it follows to do it here too. Are there values that feels more like 1, 1, 1 where still renders something, but feels less arbitrary?

Comment (txoof) [reply flat_pack_library.scad]: I'll look into that; it's possible that this is a bad idea.

File: flat_pack_library.scad (hunk 4 above)
Comment (txoof) [reply flat_pack_library.scad]: Good idea.

**Comment by txoof** [comment]
I've made the suggested changes; do I open a new PR? 

How do I best signal that this is in a state that's ready for review (again)?

**Comment by xxv** [comment]
Generally, as long as the changes are on the same topic just push them on up to this PR! 😄 I got plenty of emails signaling that it's Ready for [Re]review (RFR)

File: flat_pack_library.scad [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+Create a set of finger-joint cuts that result in two larger cuts taken at the outside
+edge
+  ###parameters:
+    *length* (real)         length of edge
+    *finger* (real)         length of each individual finger
+    *material* (real)       thickness of material - sets cut depth
+    *center* (boolean)         center the set of fingers with respect to origin
+*/
+
+module outside_cuts(length=6, finger=1, material=1, text=false, center=false, font="Liberation Sans") {
```
Comment (xxv) [inline flat_pack_library.scad]: Could the `font` and `text` parameters be removed?

File: flat_pack_library.scad [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+Create a set of finger-joint cuts that result in two larger cuts taken at the inside
+edge
+  ###parameters:
+    *length* (real)         length of edge
+    *finger* (real)         length of each individual finger
+    *material* (real)       thickness of material - sets cut depth
+    *center* (boolean)         center the set of fingers with respect to origin
+*/
+
+module inside_cuts(length=6, finger=1, material=1, text=false, center=false, font="Liberation Sans") {
```
Comment (xxv) [inline flat_pack_library.scad]: Could the `font` and `text` parameters be removed?

File: flat_pack_library.scad [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    outside_cuts_debug(length=10, center=true);
+    //outside_cuts(length=10,center = true);
+  color("red")
+    //inside_cuts(length=10,center=true);
+    inside_cuts_debug(length=10, center=true);
+}
+
+
+/*
+##module: inside_cuts
```
Comment (xxv) [inline flat_pack_library.scad]: Looks like the help text here got duplicated?

File: flat_pack_library.scad [hunk 8]
```diff
@@ -0,0 +1,170 @@
+//openscad flat-pack joint library
+
+module test() {
```
Comment (xxv) [inline flat_pack_library.scad]: Is this supposed to be here? `test` seems like a module name that would likely conflict with other modules if this library were imported.

File: flat_pack_library.scad [hunk 9]
```diff
@@ -0,0 +1,170 @@
+//openscad flat-pack joint library
```
Comment (xxv) [inline flat_pack_library.scad]: Could the name just be `flat_pack.scad`? Library should be implied.

File: flat_pack_library.scad [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  ###parameters:
+    *length* (real)         length of edge 
+    *finger* (real)         length of each individual finger
+    *material* (real)       thickness of material - sets cut depth
+    *center* (boolean)         center the set of fingers with respect to origin
+*/
+
+
+/*
+##module: intside_cuts_debug
```
Comment (xxv) [inline flat_pack_library.scad]: Just for readability's sake, could the non-debug modules go first?

Also, what tool do you use to render this help text?

File: flat_pack_library.scad (hunk 6 above)
Comment (xxv) [reply flat_pack_library.scad]: If you do `center=false`, the inside cuts don't mesh with outside cuts. Could this be aligned so that they would?

**Review by xxv — CHANGES_REQUESTED** [review]
A few more things that can help keep this nice and tidy :-)

File: flat_pack_library.scad [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+module outside_cuts(length=6, finger=1, material=1, text=false, center=false, font="Liberation Sans") {
+  // overage to ensure that all cuts are completed
+  overage = 0.0001;
+
+
+  //maximum possible divisions for this length
+  max_divisions = floor(length/finger);
+  //number of usable divisions that fall completely within the edge
+  //for this implentation the number of divisions must be odd
+  usable_divsions = max_divisions%2==0 ? max_divisions-3 : max_divisions-2;
```
Comment (xxv) [inline flat_pack_library.scad]: Whoops, just noticed a typo in `usable_divisions` here (and in the inside cuts module too)

File: flat_pack_library.scad (hunk 10 above)
Comment (txoof) [reply flat_pack_library.scad]: There's no tool; I typically just use the DaringFireball markdown so I can copy and paste it into thingiverse.
If you know a better tool I'd love to learn it. Hammering this out by hand is never pretty.

File: flat_pack_library.scad (hunk 9 above)
Comment (txoof) [reply flat_pack_library.scad]: Done!

File: flat_pack_library.scad (hunk 8 above)
Comment (txoof) [reply flat_pack_library.scad]: renamed

File: flat_pack_library.scad (hunk 10 above)
Comment (xxv) [reply flat_pack_library.scad]: Cool. I'll see what I can do to put together a simple extractor.

File: flat_pack.scad [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+Create a set of finger-joint cuts all of the same size  
+
+###parameters:
+    *length* (real)         length of edge
+    *finger* (real)         length of each individual finger
+    *material* (real)       thickness of material - sets cut depth
+    *center* (boolean)         center the set of fingers with respect to origin
+*/
+
+module inside_cuts(length=6, finger=1, material=1, text=false, center=false) {
```
Comment (xxv) [inline flat_pack.scad]: Looks like there's still the `text` parameter here and it's not being used. Same in `outside_cuts`

Comment (xxv) [reply flat_pack.scad]: In order to help with the principle of least surprise, could this module be tweaked such that when you place it un-centered at `[0, 0]` it meshes with results of `outside_cuts` with the same parameters? In my modifications, I just did a `difference()` between `outside_cuts` and the appropriate-length rectangle.

File: flat_pack.scad [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+Create a set of finger-joint cuts that result in two larger cuts taken at the outside
+edge
+  ###parameters:
+    *length* (real)         length of edge
+    *finger* (real)         length of each individual finger
+    *material* (real)       thickness of material - sets cut depth
+    *center* (boolean)         center the set of fingers with respect to origin
+*/
+
+module outside_cuts(length=6, finger=1, material=1, text=false, center=false) {
```
Comment (xxv) [inline flat_pack.scad]: unused `text` parameter.

File: flat_pack.scad [hunk 14]
```diff
@@ -0,0 +1,169 @@
+//openscad flat-pack joint library
```
Comment (xxv) [inline flat_pack.scad:1]: I'm not sure what tool you use to edit this file, but one small nit-picky detail that my tool (vim) shows is that some lines have stray whitespace at the end of them. It's not a big deal, but just a little thing that keeps your code cleaner and is generally good practice to look out for 😄

File: flat_pack.scad [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+module finger_testing() {
```
Comment (xxv) [inline flat_pack.scad]: Could this just go in a separate file that demos the way this library work? Maybe shows how all the parameters can be varied and how you can make a mating joint using `centered=true` vs `centered=false`.

**Review by xxv — CHANGES_REQUESTED** [review]
Almost there! Home stretch!

**Review by txoof — COMMENTED** [review]
I think we're ready to merge this into the main branch. Can you give me some advice as to how to do this?

If there are any more changes, let me know.

**Comment by xxv** [comment]
Did you see my recent review comments? There are just a few small things and then it should be ready for merge.

File: flat_pack.scad (hunk 14 above)
Comment (txoof) [reply flat_pack.scad:1]: I'm using vim too; What's your favorite way to visualize unneeded white space? I never really thought about this before. 
I found this useful: http://vim.wikia.com/wiki/Highlight_unwanted_spaces

**Comment by txoof** [comment]
I missed your last review; but I think I've covered all the suggestions you made. Nearly there!

File: flat_pack.scad [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+module outside_cuts(length=6, finger=1, material=1, center=false) {
+  // overage to ensure that all cuts are completed
+  overage = 0.0001;
+
+
+  //maximum possible divisions for this length
+  max_divisions = floor(length/finger);
+  //number of usable divisions that fall completely within the edge
+  //for this implentation the number of divisions must be odd
```
Comment (xxv) [inline flat_pack.scad]: Same typo as the above bit, "implementation"

BTW: in vim I use

```
set spell
```

along with

```
syntax on
```

and other such things. That will help find typos pretty easily. Here's my full, ugly vimrc:

http://xxv.so/0x93653a40

File: flat_pack_demo.scad [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+module finger_testing() {
+  difference() {
+    color("blue")
+      square([110, 40]);
+      outside_cuts(length=110, finger=5, material=10, center=false);
+  }
+
+  rotate([180, 0, 0])
+  translate([0, 0])
```
Comment (xxv) [inline flat_pack_demo.scad]: Can probably remove the `translate([0, 0])` here and below.

File: flat_pack.scad [hunk 18]
```diff
@@ -0,0 +1,153 @@
+//openscad flat-pack joint library
+
+/*
+##module: outside_cuts
+Create a set of finger-joint cuts that result in two larger cuts taken at the outside
+edge
+  ###parameters:
```
Comment (xxv) [inline flat_pack.scad]: Put a blank line above this one and shift this left? I think that's how markdown works.

File: flat_pack.scad [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        square([finger, material+overage]); //add a small amount to ensure complete cuts
+    }
+  }
+
+} //end inside_cuts
+
+/*
+##module: intside_cuts_debug
+Create a set of finger-joint cuts all of the same size
+  ###parameters:
```
Comment (xxv) [inline flat_pack.scad]: Same header shift here, I think. May be wise to check in a markdown parser.

File: flat_pack.scad [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+
+} // end inside debug
+
+
+/*
+##module: outside_cuts_debug
+Create a set of finger-joint cuts that result in two larger cuts taken at the i
+inside edge with debugging text
+  ###parameters:
```
Comment (xxv) [inline flat_pack.scad]: Same indentation/whitespace issue as above

**Review by xxv — APPROVED** [review]
Ok! Looks like all the core functionality is there. Please see the various small review comments. They're mostly little details involving formatting.

I'll leave you a message on IRC for details on merging.

File: flat_pack.scad [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+
+    // make both endcuts here
+    for (j = [0, 1]) {
+      translate([(length-end_cut_length)*j, -overage/2])
+        square([end_cut_length, material+overage]);
+    }
+
+
+    //make all the "normal" finger cuts here
+    for (i = [0 : num_cuts-1]) {
```
Comment (xxv) [inline flat_pack.scad]: I had to wrap this `for` statement in:

```
if (num_cuts > 0) {
  for (i = [0 : num_cuts-1]) {
// ...
```

as OpenSCAD was giving me errors when num_cuts was 0.

**Review by xxv — CHANGES_REQUESTED** [review]
I got a warning in OpenSCAD when running my boxmaker code on this library.

File: flat_pack.scad (hunk 21 above)
Comment (xxv) [reply flat_pack.scad]: Same is true with the similar statement in the other module.

Comment (txoof) [reply flat_pack.scad]: The if() solution works, but does it make sense to catch that? Shouldn't we get a warning when the number of cuts is 0? 
In what scenarios is it OK to have 0 cuts when calling a module that makes cuts?

Comment (txoof) [reply flat_pack.scad]: After thinking more about this, you're right. OpenSCAD doesn't like iterating over ranges that are negative and when num_cuts is 0, the for loop goes from `[0 : -1]` and that's not correct.

**Pull request closed without merging**